File: observer.py
from abc import ABC, abstractmethod


class Observador(ABC):
    @abstractmethod
    def actualizar(self, temperatura):
        pass


class SensorTemperatura:
    def __init__(self):
        self._observadores = []
        self._temperatura = None

    def agregar_observador(self, observador):
        self._observadores.append(observador)

    def eliminar_observador(self, observador):
        self._observadores.remove(observador)

    def notificar(self):
        for obs in self._observadores:
            obs.actualizar(self._temperatura)

    def set_temperatura(self, nueva_temp):
        print(f"\n[Sensor] Nueva temperatura detectada: {nueva_temp}°C")
        self._temperatura = nueva_temp
        self.notificar()


class Pantalla:
    def actualizar(self, temperatura):
        print(f"[Pantalla] Temperatura actual: {temperatura}°C")

class Alarma:
    def actualizar(self, temperatura):
        if temperatura > 50:
            print(f"[Alarma] ¡Temperatura peligrosa!: {temperatura}°C")
        else:
            print(f"[Alarma] Temperatura normal")

class RegistroBD:
    def actualizar(self, temperatura):
        print(f"[Base de Datos] Registro guardado: {temperatura}°C")


sensor = SensorTemperatura()

pantalla = Pantalla()
alarma = Alarma()
bd = RegistroBD()

sensor.agregar_observador(pantalla)
sensor.agregar_observador(alarma)
sensor.agregar_observador(bd)


sensor.set_temperatura(30)
sensor.set_temperatura(55)

File: java/observer.java
import java.util.ArrayList;
import java.util.List;

interface Observador {
    void actualizar(int temperatura);
}

class SensorTemperatura {
    private List<Observador> observadores = new ArrayList<>();
    private int temperatura;

    public void agregarObservador(Observador obs) {
        observadores.add(obs);
    }

    public void eliminarObservador(Observador obs) {
        observadores.remove(obs);
    }

    public void notificar() {
        for (Observador obs : observadores) {
            obs.actualizar(temperatura);
        }
    }

    public void setTemperatura(int nuevaTemp) {
        System.out.println("\n[Sensor] Nueva temperatura detectada: " + nuevaTemp + "°C");
        this.temperatura = nuevaTemp;
        notificar();
    }
}

class Pantalla implements Observador {
    @Override
    public void actualizar(int temperatura) {
        System.out.println("[Pantalla] Temperatura actual: " + temperatura + "°C");
    }
}

class Alarma implements Observador {
    @Override
    public void actualizar(int temperatura) {
        if (temperatura > 50) {
            System.out.println("[Alarma] ¡Temperatura peligrosa!: " + temperatura + "°C");
        } else {
            System.out.println("[Alarma] Temperatura normal");
        }
    }
}

class RegistroBD implements Observador {
    @Override
    public void actualizar(int temperatura) {
        System.out.println("[Base de Datos] Registro guardado: " + temperatura + "°C");
    }
}

public class observer {
    public static void main(String[] args) {
        SensorTemperatura sensor = new SensorTemperatura();

        Pantalla pantalla = new Pantalla();
        Alarma alarma = new Alarma();
        RegistroBD bd = new RegistroBD();

        sensor.agregarObservador(pantalla);
        sensor.agregarObservador(alarma);
        sensor.agregarObservador(bd);

        sensor.setTemperatura(30);
        sensor.setTemperatura(55);
    }
}
